docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,226 @@
+This structure is used by both "from" and "to" policies plugins.
+Policy plugins use the `Rules` list to compute `Conf` for the known destination.
+In example, if we want to know the `Conf` for outbound with `kuma.io/service: bar` tags,
+we're going to iterate over `Rules` and check if our outbound belongs to the `Rules[].Subset`.
+
+The approach has a number of disadvantages:
+
+* even though it works well for "from" policies, using it for "to" policies doesn't feel natural
+  * there is no concept of "subsets" in outbounds, we have to artificially build `{kuma.io/service: $svc}` subset
+  * subset-based matching is build to handle subset overlapping scenarios, but it's never happening in outbounds
+* when referencing the real MeshHTTPRoute we had to use subset with artificial `__rule-matches-hash__` key
+* adding support for `MeshService` forces us to introduce another artificial key like `__resource-name__`
+* Inspect API with subsets is tricky when referencing real resources (like MeshHTTPRoute or MeshService)
+
+We need a new format that'd make things easier when it comes to referencing real resources in `to[]`.
+
+## Considered Options and Decision Outcome
+
+No changes for "from" policies, both policy matching algorithm and Inspect API stays the same.
+
+For "to" policies the algorithm is going to return the additional structure:
+
+```go
+type ToRules struct {
+    Rules         Rules 
+    ResourceRules map[UniqueResourceKey]ResourceRule // new field
+}
+
+// UniqueResourceKey is a way to uniquely identify a resource, should include ResourceType, Name and Mesh
+type UniqueResourceKey struct{}
+
+type ResourceRule struct {
+    Resource core_model.ResourceMeta
+    Conf     interface{}
+    Origin   []Origin
+}
+
+type Origin struct {
+    Resource core_model.ResourceMeta
+    // ToIdx is an index in the 'to[]' array, so we could unambiguously detect what to-item contributed to the final conf.
+    // Especially useful when to-item uses `targetRef.Labels`, because there is no obvious matching between the specific resource
+    // in `ResourceRule.Resource` and to-item.
+    ToIdx int 
+}
+```
+
+For a simple targetRef with `name/namespace`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+      namespace: finance
+    conf: conf1
+```
+
+we're going to produce a single `ResourceRule` entity:
+
+```yaml
+resourceRules:
+  meshservice.mesh-1.backend.finance:
+    resource:
+      name: backend.finance
+      mesh: mesh-1
+    conf: conf-1
+    origin: [...]
+```
+
+When targetRef has kind `MeshService` and uses `labels` we're going to produce multiple `ResourceRule` entities
+for each real `MeshService` that satisfies `labels`, i.e for the given `targetRef`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      labels:
+        k8s.kuma.io/namespace: finance
+    conf: conf1
+```
+
+we are going to fetch all `MeshService` resources, filter them by `k8s.kuma.io/namespace: finance` and produce N rules:
```

Comment:
What's the merging story?


What happens when I have:

```yaml
type: MeshService
name: my-service
labels:
  a-common-label: foo
---
type: MeshService
name: labeled-service
labels:
  a-common-label: foo
---
type: MeshService
name: other-service
labels: {}
---
kind: Policy
spec:
  targetRef:
    kind: Mesh
  to:
    - targetRef:
         kind: Mesh
       conf: $conf1
---
kind: Policy
spec:
  targetRef:
    kind: Mesh
  to:
    - targetRef:
         kind: MeshService
         labels:
             a-common-label: foo
       conf: $conf2
---
kind: Policy
spec:
  targetRef:
    kind: Mesh
  to:
    - targetRef:
         kind: MeshService
         name: my-service
       conf: $conf3
```
I'd expect:

```
resourceRules:
   my-service:
      conf: merge($conf1, $conf2, conf3)
   labeled-service:
     conf: merge($conf1, $conf2)
   other-service:
     conf: merge($conf1)
```

Is this a correct understanding? If it is it should be explicitly stated.

Also what about Mesh*Route on top of this?

Reply:
That's correct understanding with one caveat – at this moment the order of confs inside `merge()` is defined by the alphabetical order of `metadata.name`. So depending on the alphabetical order you can get either:
```
my-service:
  conf: merge($conf1, $conf2, $conf3)
```
or 
```
my-service:
  conf: merge($conf2, $conf3, $conf1)
```
and etc.

That's the same thing you mentioned here https://github.com/kumahq/kuma/issues/9151.

Reply:
> Also what about Mesh*Route on top of this?

I'm planning to implement function `includes(r1, r2 core_model.Resource) bool` that returns true if 1 resource is included into another:
```
* Mesh includes every other resource
* MeshService includes MeshHTTPRoutes and MeshTCPRoute
```
it's already implemented here for simple cases https://github.com/kumahq/kuma/pull/10886/files#diff-b78b017a9fd32d0715c7c584478b3836fd22ae71c3bc2b49eb06480e3866d5b8R80.

I'll update MADR to reflect that.

Reply:
Added information about merging